proxy/coco/src/peer.rs

```diff
@@ -48,21 +48,60 @@ pub enum Error {
     #[error(transparent)]
     Announcement(#[from] announcement::Error),
 
-    /// There was an error in a spawned task.
+    /// Peer bootstrap error.
+    #[error(transparent)]
+    Bootstrap(#[from] net::protocol::error::Bootstrap),
+
+    /// Encountered an I/O error, for example fetching the peer's listen addresses.
+    #[error(transparent)]
+    Io(#[from] io::Error),
+
+    /// The joining of a thread failed.
     #[error("the running peer was either cancelled, or one of its tasks panicked")]
-    Spawn(#[source] spawn_abortable::Error),
+    Join(#[source] JoinError),
 
-    /// There was an error when interacting with [`State`].
+    /// The protocol encountered an error.
     #[error(transparent)]
-    State(#[from] state::Error),
+    Protocol(#[from] net::quic::Error),
+
+    /// An interaction with the underlying storage failed.
+    #[error(transparent)]
+    State(#[from] state::error::Error),
+}
+
+/// Constructs a [`Peer`] from a [`net::peer::Config`].
+///
+/// # Errors
+///
+/// * peer construction from config fails.
+/// * accept on the peer fails.
+pub async fn bootstrap<D>(
+    config: net::peer::Config<BoxedSigner>,
+    disco: D,
+    store: kv::Store,
+    run_config: RunConfig,
+) -> Result<Peer<D>, Error>
+where
+    D: net::discovery::Discovery<Addr = SocketAddr> + Send + 'static,
+    <D as net::discovery::Discovery>::Stream: 'static,
+{
+    let peer = librad::net::peer::Peer::new(config);
+    let bound = peer.bind().await?;
+
+    let peer = Peer::new(peer, bound, disco, store, run_config)?;
+
+    Ok(peer)
 }
 
 /// Local peer to participate in the radicle code-collaboration network.
-pub struct Peer {
-    /// Peer [`librad::net::peer::RunLoop`] to advance the network protocol.
-    run_loop: RunLoop,
-    /// Underlying state that is passed to subroutines.
-    state: State,
```

Comment:
`net::peer::Peer` is the new `State`. It's essentially what was the `PeerApi`, but a simplified and less footgun-y version